### PR TITLE
xds: Ext proc client application level flow control

### PR DESCRIPTION
**Pull request by kannanjgithub**

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 1]
```diff
@@ -278,18 +274,66 @@ private static class DataPlaneClientCall
...
+    @GuardedBy("streamLock")
+    private final Queue<ByteString> pendingUpstreamBodyMessages =
+        new java.util.concurrent.ConcurrentLinkedQueue<>();
+    // Path 4: Outstanding requests from downstream for pulling responses
+    @GuardedBy("streamLock")
+    private int downstreamRequestsPending = 0;
+    // Buffered mutated response bodies from ext_proc server
+    @GuardedBy("streamLock")
+    private final Queue<ByteString> pendingMutatedResponseBodies =
+        new java.util.concurrent.ConcurrentLinkedQueue<>();
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: nit: Remove the fully qualified import

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 2]
```diff
@@ -439,6 +479,25 @@ public void onNext(ProcessingResponse response) {
               return;
             }
 
+            if (response.hasServerWindowUpdate()) {
+              ProcessingResponse.ServerWindowUpdate update = response.getServerWindowUpdate();
+              boolean wasReady = isReady();
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Why does this happen outside the streamlock? 
This has potential to cause hangs due to lock acquisition and releases. 

If an application thread consumes remaining window credit between the first `isReady` check and the synchronized block, the final  check evaluates to false and drops the onReadyNotify() which can result in hangs?

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 3]
```diff
@@ -667,21 +726,94 @@ private void sendToExtProc(ProcessingRequest request) {
           requestToSend = ProcessingRequest.newBuilder(requestToSend)
               .setObservabilityMode(true)
               .build();
+        } else if (!flowControlInitSent) {
+          requestToSend = ProcessingRequest.newBuilder(requestToSend)
+              .setFlowControlInit(ProcessingRequest.FlowControlInit.newBuilder()
+                  .setInitialWindowDownstreamToSidestream(DEFAULT_INITIAL_WINDOW_SIZE)
+                  .setInitialWindowSidestreamToUpstream(DEFAULT_INITIAL_WINDOW_SIZE)
+                  .setInitialWindowUpstreamToSidestreama(DEFAULT_INITIAL_WINDOW_SIZE)
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: nit : type : `setInitialWindowUpstreamToSidestreama` -> `setInitialWindowUpstreamToSidestream`.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 4]
```diff
@@ -667,21 +726,94 @@ private void sendToExtProc(ProcessingRequest request) {
...
+      long incrementUpstream = accumulatedWindowUpdateSidestreamToUpstream;
+      long incrementDownstream = accumulatedWindowUpdateSidestreamToDownstream;
+
+      if (incrementUpstream > 0 || incrementDownstream > 0) {
+        requestBuilder.setClientWindowUpdate(
+            ProcessingRequest.ClientWindowUpdate.newBuilder()
+                .setWindowIncrementSidestreamToUpstream(incrementUpstream)
+                .setWindowIncrementSidestreamToDownstream(incrementDownstream)
+                .build());
+        accumulatedWindowUpdateSidestreamToUpstream -= incrementUpstream;
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:764]: Optional: 
The side effect here seems somewhat odd. We are mutating the accumulation while building the request. This assumes that we expect a guarantee that the request will always be build and sent. If some implementation in future adds a conditional around this, we may end up in a scenario where we performed the mutation on accumulation but never sent the update.

While the currently, I can't think of a case where this'd reasonably affect anything, but I am slightly afraid of maintaining this invariant , especially when undocumented.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 5]
```diff
@@ -812,29 +1004,58 @@ public void sendMessage(InputStream message) {
...
-        }
-      } catch (IOException e) {
-        rawCall.cancel("Failed to serialize message for External Processor", e);
+    @GuardedBy("streamLock")
+    private void drainPendingRequestBodyMessages() {
+      while (downstreamToSidestreamWindow > 0 && !pendingRequestBodyMessages.isEmpty()) {
+        ByteString body = pendingRequestBodyMessages.poll();
+        sendRequestBodyToExtProc(body);
+      }
+      if (pendingRequestBodyMessages.isEmpty() && pendingHalfClose.get()) {
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: I think this should be a compareandSet to ensure that we call `halfClose` only once. 

Alternatively we can make halfClose idempotent, if there are multiple callsites.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 6]
```diff
@@ -914,11 +1142,31 @@ private void handleRequestBodyResponse(BodyResponse bodyResponse) {
         if (mutation.hasStreamedResponse()) {
           StreamedBodyResponse streamed = mutation.getStreamedResponse();
           if (!streamed.getEndOfStreamWithoutMessage()) {
-            super.sendMessage(new KnownLengthInputStream(streamed.getBody()));
+            com.google.protobuf.ByteString body = streamed.getBody();
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: nit: I believe we have an import for this.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 7]
```diff
@@ -914,11 +1142,31 @@ private void handleRequestBodyResponse(BodyResponse bodyResponse) {
...
+            boolean sendImmediately = false;
+            synchronized (streamLock) {
+              sidestreamToUpstreamWindow -= body.size();
+              if (pendingUpstreamBodyMessages.isEmpty() && super.isReady()) {
+                sendImmediately = true;
+                accumulatedWindowUpdateSidestreamToUpstream += body.size();
+              } else {
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1167]: We may have started reaching high levels of cyclomatic complexity. This seems to be nearly 4 levels of conditional nesting. We should consider refactoring to simplify things , not necessarily in this PR though.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 8]
```diff
@@ -914,11 +1142,31 @@ private void handleRequestBodyResponse(BodyResponse bodyResponse) {
         if (mutation.hasStreamedResponse()) {
           StreamedBodyResponse streamed = mutation.getStreamedResponse();
           if (!streamed.getEndOfStreamWithoutMessage()) {
-            super.sendMessage(new KnownLengthInputStream(streamed.getBody()));
+            com.google.protobuf.ByteString body = streamed.getBody();
+            boolean sendImmediately = false;
+            synchronized (streamLock) {
+              sidestreamToUpstreamWindow -= body.size();
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1163]: We seem to be unconditionally subtracting things here. I am guessing this relies on the assumption that the sender respects the flow control behavior.

I looked at the grfc and couldn't find any guidelines for offending server. I also realized that one of the commits explicitly changed this behavior. 

This in the current state feels dangerous in the sense that it may result in OOMs. Do we have any protection? Should this be discussed in the grfc?

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 9]
```diff
@@ -768,16 +933,43 @@ public void request(int numMessages) {
...
         return;
       }
-      if (!isSidecarReady()) {
-        pendingRequests.addAndGet(numMessages);
-        return;
+      synchronized (streamLock) {
+        boolean sendResponseBodiesToExtProc = config.getObservabilityMode()
+            || currentProcessingMode.getResponseBodyMode() == ProcessingMode.BodySendMode.GRPC;
+
+        if (!sendResponseBodiesToExtProc) {
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: This is likely dead code, since we have a similar check above.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 10]
```diff
@@ -1145,34 +1524,60 @@ public void onMessage(InputStream message) {
...
-        dataPlaneClientCall.bodyMessageSentToExtProc.set(true);
+        try {
+          ByteString bodyByteString = ByteString.readFrom(message);
+          if (dataPlaneClientCall.getConfig().getObservabilityMode()) {
+            sendResponseBodyToExtProc(bodyByteString, false);
+            dataPlaneClientCall.bodyMessageSentToExtProc.set(true);
+            dataPlaneClientCall.getCallContext().run(
+                () -> delegate().onMessage(bodyByteString.newInput()));
+          } else {
+            if (dataPlaneClientCall.upstreamToSidestreamWindow <= 0 || !savedMessages.isEmpty()) {
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1567]: Should we be mutating the private data members of dataPlaneClientCall directly or use encapsulation here and at other places?

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 11]
```diff
@@ -1145,34 +1524,60 @@ public void onMessage(InputStream message) {
...
+          return;
+        }
 
-      try {
-        ByteString bodyByteString = ByteString.readFrom(message);
-        sendResponseBodyToExtProc(bodyByteString, false);
-        dataPlaneClientCall.bodyMessageSentToExtProc.set(true);
+        try {
+          ByteString bodyByteString = ByteString.readFrom(message);
+          if (dataPlaneClientCall.getConfig().getObservabilityMode()) {
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1561]: Optional: Now that flow control differs across observability and normal mode. Does it make sense to revisit the decision to separate them via polymorphism instead of procedural conditions? Not necessarily in this PR.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 12]
```diff
@@ -931,7 +1179,102 @@ private void handleResponseBodyResponse(
         BodyMutation mutation = bodyResponse.getResponse().getBodyMutation();
         if (mutation.hasStreamedResponse()) {
           StreamedBodyResponse streamed = mutation.getStreamedResponse();
-          listener.onExternalBody(streamed.getBody());
+          com.google.protobuf.ByteString body = streamed.getBody();
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: nit: Fully qualified type

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 13]
```diff
@@ -931,7 +1179,102 @@ private void handleResponseBodyResponse(
...
+          final int bodySize = body.size();
+          synchronized (streamLock) {
+            sidestreamToDownstreamWindow -= bodySize;
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1200]: Same as the other comment. Is the unconditional subtraction expected?

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 14]
```diff
@@ -931,7 +1179,102 @@ private void handleResponseBodyResponse(
...
+      }
+    }
+
+    private void drainPendingMutatedResponseBodies() {
+      List<ByteString> toDeliver = new ArrayList<>();
+      synchronized (streamLock) {
+        while (downstreamRequestsPending > 0 && !pendingMutatedResponseBodies.isEmpty()) {
+          ByteString body = pendingMutatedResponseBodies.poll();
+          downstreamRequestsPending--;
+          pendingRequests.decrementAndGet();
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1238]: It's difficult to follow why pendingRequests need to be decremented here. I believe this is due to the fact it's only called from `request(..)`. 
The method below makes things more confusing is the existence of similarly named `drainPendingMutatedResponseBodiesDirect` . I think we may want to distinguish between them via comments.

**Comment by sauravzg** [comment]
First pass at source done. Mostly okay. I have some concerns about the unbounded receiver buffering and some non-urgent concerns about growing complexity and size of the file, but the latter can potentially be addressed in the future, since it'll likely be a large refactor.

Will look at the tests later.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 1 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Done.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 2 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Good catch, fixed it now.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 3 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Done.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 4 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:764]: I considered changing
```
void mergeAccumulatedWindowUpdates(ProcessingRequest.Builder requestBuilder)
```
to
```
ProcessingRequest mergeAccumulatedWindowUpdates(boolean isRequest, ByteString body, boolean isEndOfStream)
```
but one is still free to throw away the return value even though calling the method changes the value of the state variables. I don't seem to have a real solution to this while having readable code with methods for specific work items.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 9 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Removed.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 5 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Good catch. Added compareandSet to ensure that we call halfClose only once.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 6 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Done.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 8 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1163]: As discussed offline, this was deemed inevitable with allowing negative window updates.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 7 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1167]: Will think about it.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 12 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java]: Done.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 13 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1200]: Yes.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 14 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1238]: `pendingRequests` tracks the number of messages we need to request from the backend server transport (super.request()) but have delayed due to flow control constraints. `downstreamRequestsPending` tracks the number of messages the client application has requested from the interceptor but has not yet received. When the app calls `request()`, we first try to satisfy this demand using the buffered mutated messages. For each buffered message we deliver, we decrement `downstreamRequestsPending` and also decrement `pendingRequests` (since we satisfied the app without needing to pull a new message from the backend).

`drainPendingMutatedResponseBodiesDirect` is to immediately flush any mutated response chunks that we already received and buffered before the stream failed, ensuring the application receives them in the correct order.
Since the counters (`downstreamRequestsPending` and `pendingRequests`) are completely ignored once the pass-through mode is active, there is no need to update them during this final flush. Added a comment for `drainPendingMutatedResponseBodiesDirect`.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 11 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1561]: Added a TODO to consider doing it.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 10 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1567]: It is not against style guidelines to mutate private data members of private classes.

File: xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java [hunk 15]
```diff
@@ -9988,1133 +10093,6035 @@ public void onError(Throwable t) {
...
+
+    assertThat(sidecarLatch.await(5, TimeUnit.SECONDS)).isTrue();
+
+    assertThat(receivedRequests).hasSize(2);
+    ProcessingRequest firstRequest = receivedRequests.get(0);
+    ProcessingRequest secondRequest = receivedRequests.get(1);
+
+    assertThat(firstRequest.hasRequestHeaders()).isTrue();
+    assertThat(firstRequest.hasFlowControlInit()).isTrue();
+    assertThat(firstRequest.getFlowControlInit().getInitialWindowDownstreamToSidestream())
```
Comment (sauravzg) [inline xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java:12143]: Should we be covering all 4 windows here?

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 4 above)
Comment (sauravzg) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:764]: Fair enough. Let's document it as function documentation and that should be good enough.

File: xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java
Comment (sauravzg) [inline xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java:1]: Let's add unit tests for the fixes made in the latest commit for addressing review comments as applicable.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java [hunk 16]
```diff
@@ -1145,34 +1520,61 @@ public void onMessage(InputStream message) {
...
 
-        if (dataPlaneClientCall.getConfig().getObservabilityMode()) {
-          // If needed, downstream reading can be made more optimal by creating a wrapped
-          // Inputstream wraps the underlying bytestring and that implements HasByteBuffer,
-          // Detachable, KnownLength
-          dataPlaneClientCall.getCallContext().run(
-              () -> delegate().onMessage(bodyByteString.newInput()));
+    void drainSavedMessages() {
+      synchronized (dataPlaneClientCall.streamLock) {
+        while (dataPlaneClientCall.isSidecarReady()
```
Comment (sauravzg) [inline xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1584]: IDE shows some missed coverage here. 

My likely guess is that we don't have an exit criteria out of the ` dataPlaneClientCall.upstreamToSidestreamWindow > 0` i.e. we need behavorial tests that cover cases where the window update doesn't empty the entire queue.

Might be worth adding.

**Comment by sauravzg** [comment]
Looked at the tests. Things seem reasonable. Unfortunately the CI are failing. I'll be OOO for most of next weeek, but feel free to ping me once the remaining comments are addresses and the CI is passing. I'll gladly approve.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 8 above)
Comment (sauravzg) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1163]: So, while this is compliant with the grfc and the reason grfc chose to do this also seems reasonable, I think we should still follow up on the grfc to understand the expected behavior from the client side against an offending server.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 4 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:764]: Added a comment that the method has side-effects.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 8 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1163]: I have raised a question in the grfc.

File: xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java
Comment (kannanjgithub) [reply xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java:1]: For the `pendingHalfClose` race, I have added `testNoDuplicateHalfCloseSentToExtProc`. I have refactored the half-close state management to use a new `appHalfClosed` flag (tracking whether the application has requested a half-close) and kept `pendingHalfClose` strictly as a transient flag for whether we need to send the half-close request to the ext_proc server when the message queue becomes empty.

This decouples the two states and prevents `drainPendingRequestBodyMessages()` from repeatedly triggering `halfClose()` (which was resetting `pendingHalfClose` to true internally inside `halfClose()`).

Regarding the `isReady()` race condition fix it is not possible to write a reliable, non-flaky unit test for this.

File: xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java (hunk 16 above)
Comment (kannanjgithub) [reply xds/src/main/java/io/grpc/xds/ExternalProcessorClientInterceptor.java:1584]: Added test `testUpstreamToSidestreamFlowControl_DrainsPartiallyOnPartialWindowReplenishment`.

**Review by sauravzg — APPROVED** [review]
There's probably a small minor comment remaining and the CI still needs to be fixed. Approving, please take care of these before submitting

File: xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java (hunk 15 above)
Comment (kannanjgithub) [reply xds/src/test/java/io/grpc/xds/ExternalProcessorClientInterceptorTest.java:12143]: Done.
